src/realm/object-store/sectioned_results.cpp

```diff
@@ -0,0 +1,275 @@
+////////////////////////////////////////////////////////////////////////////
+//
+// Copyright 2022 Realm Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+
+#include <realm/object-store/results.hpp>
+#include <realm/object-store/sectioned_results.hpp>
+#include <realm/object-store/thread_safe_reference.hpp>
+
+namespace realm {
+
+// This method will run in the following scenarios:
+// - SectionedResults is freshly created from Results
+// - The underlying Table in the Results collection has changed
+static std::vector<SectionRange> calculate_sections(Results& results,
+                                                    const SectionedResults::SectionKeyFunc& callback)
+{
+    auto sections = std::map<Mixed, SectionRange>();
+    // Take a snapshot in case the underlying results change while
+    // the calculation is being performed.
+    auto snapshot = results.snapshot();
+    const size_t size = snapshot.size();
+    size_t current_section = 0;
+    for (size_t i = 0; i < size; ++i) {
+        Mixed section_key = callback(snapshot.get_any(i), snapshot.get_realm());
+        // Disallow links as section keys. It would be uncommon to use them to begin with
+        // and if the object acting as the key was deleted bad things would happen.
+        if (!section_key.is_null() && section_key.get_type() == type_Link) {
+            throw std::logic_error("Links are not supported as section keys.");
+        }
+        if (sections.find(section_key) == sections.end()) {
+            SectionRange section;
+            section.index = current_section;
+            section.key = section_key;
+            section.indices.push_back(i);
+            sections[section_key] = section;
+            current_section++;
+        }
+        else {
+            sections[section_key].indices.push_back(i);
+        }
+    }
+
+    std::vector<SectionRange> offset_ranges;
```

Comment:
Rather than copying the data from the temporary map to a vector, the actual data could be stored in the vector from the beginning with the map being key -> index in vector.